starboard/shared/starboard/media/mime_util.cc

```diff
@@ -99,9 +99,20 @@ bool IsSupportedAudioCodec(const ParsedMimeInfo& mime_info) {
       break;
 #if SB_API_VERSION >= 14
     case kSbMediaAudioCodecMp3:
+      if (mime_type.subtype() != "mpeg" && mime_type.subtype() != "mp3") {
+        return false;
+      }
+      break;
     case kSbMediaAudioCodecFlac:
+      if (mime_type.subtype() != "ogg" && mime_type.subtype() != "flac") {
```

Comment:
This implies that if a Starboard implementation supports kSbMediaAudioCodecMp3, it should be able to play mp3 contained in mpeg containers, without the FFmpeg demuxer extension.  And this is implied for all the codecs and containers added in this CL.

Just want to double check if this is true?

Reply:
Without the demuxer extension or some other demuxer capable of parsing the container, no, I don't think the content will actually be played.

High-level question: why is the media container relevant to Starboard? The actual demuxing is done by Cobalt (or Cast, or something else upstream of Starboard), not Starboard itself, right? It looks like Cobalt already checks for unsupported containers here: https://github.com/youtube/cobalt/blob/022e8b98727130fe9746ab48b00188f2856d04b0/cobalt/media/media_module.cc#L124
I didn't see that code running when I tested, though, so maybe I'm missing something (is SbMediaCanPlayMimeAndKeySystem actually used by Cobalt?)

As an alternative: would it make sense to just remove the container checks here entirely, since that part doesn't seem relevant to Starboard itself? Or am I just misunderstanding?

Reply:
The check of container can (and should) happen in Cobalt.  It's now handled in Starboard for legacy reasons.  It makes sense to move it from Starboard to Cobalt, but we shouldn't just remove it in Starboard without adding proper support in Cobalt.

The checked quoted above is for progressive playback only, and is insufficient to filter out unsupported containers for adaptive playbacks (i.e. MSE based).

SbMediaCanPlayMimeAndKeySystem() is called inside MediaSource.isTypeSupported(), which can be called by the web app.  If the web app chooses to query for `audio/mpeg, codecs="mp3"` with the new change, Cobalt may incorrectly report the support.

Reply:
That makes sense; thanks for the explanation!

In that case, what if we just add a check in CanPlayAdaptive similar to the one in CanPlayProgressive? In this case, we would check that the MIME type begins with video/mp4 or video/webm. That will immediately rule out all the codecs I added (so MediaSource.isTypeSupported() will correctly return false). Based on IsSupportedVideoCodec (in mime_util.cc), webm and mp4 are the only containers supported by Cobalt.

If that solution is acceptable, let me know and I'll send out a PR.